File: api/src/main/java/net/blueshell/api/dto/survey/SurveyDTO.java
package net.blueshell.api.dto.survey;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import lombok.Data;
import lombok.EqualsAndHashCode;
import net.blueshell.api.base.BaseDTO;
import net.blueshell.api.validation.survey.ValidQuestionList;

import java.util.Comparator;
import java.util.List;

@Data
@EqualsAndHashCode(callSuper = true)
@Schema(name = "Survey")
public class SurveyDTO extends BaseDTO {
    private Long id;
    private Long responseCount;

    @NotEmpty
    @ValidQuestionList
    @Valid
    List<QuestionDTO> questions;

    @JsonProperty("questions")
    public List<QuestionDTO> getQuestionsSorted() {
        if (questions == null) return List.of();
        return questions.stream()
                .sorted(Comparator
                        .comparing(QuestionDTO::getIdx, Comparator.nullsLast(Comparator.naturalOrder()))
                ).toList();
    }
}

File: api/src/main/kotlin/net/blueshell/api/dto/survey/SurveyDTO.kt
package net.blueshell.api.dto.survey

import com.fasterxml.jackson.annotation.JsonProperty
import io.swagger.v3.oas.annotations.media.Schema
import jakarta.validation.Valid
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull
import net.blueshell.api.base.BaseDTO
import net.blueshell.api.validation.survey.ValidQuestionList
import java.util.function.Function
@Schema(name = "Survey")
class SurveyDTO : BaseDTO() {
    @NotEmpty
    @ValidQuestionList
    @Valid
    var questions: @NotEmpty @Valid MutableList<QuestionDTO?>? = null
    val responseCount: Long? = null

    @get:JsonProperty("questions")
    val questionsSorted: MutableList<QuestionDTO?>
        get() {
            if (questions == null) return mutableListOf<QuestionDTO?>()
            return questions!!.stream()
                .sorted(
                    Comparator
                        .comparing<QuestionDTO?, @NotNull Long?>(
                            Function { obj: QuestionDTO? -> obj!!.idx },
                            Comparator.nullsLast<@NotNull Long?>(Comparator.naturalOrder<@NotNull Long?>())
                        )
                ).toList()
        }
}
